### PR TITLE
LPS-148133  Stabilize BalloonEditor#CanFormatText

### PR DESCRIPTION
**Pull request by mateusralv**
**Background:**
Investigate flake fail and stabilize BalloonEditor#CanFormatText 

**JIRA Ticket:**
https://issues.liferay.com/browse/LPS-148133

### CODE REVIEW

**Comment by mateusralv** [comment]
ci:test:frontend:nocompile

**Comment by john-co** [comment]
What did you find from investigation? Was page refresh the only way to stabilize?

**Comment by mateusralv** [comment]
> ci:test:frontend:nocompile

Hi John, sorry for the delay. The failures happened because the portlet added to the page did not load at the beginning of the test. This happened previously with [FrontendDataSet#ActionsCanBeCustomized ](https://issues.liferay.com/browse/LRQA-74339)and the addition of Refresh(); solved the problem.

**Comment by mateusralv** [comment]
Hi @john-co, sorry for the delay. The failures happened because the portlet added to the page did not load at the beginning of the test. This happened previously with [FrontendDataSet#ActionsCanBeCustomized ](https://issues.liferay.com/browse/LRQA-74339)and the addition of Refresh(); solved the problem.

**Comment by john-co** [comment]
Thanks Mateus, I guess I was looking to see if you looked at other options before settling on Refresh(). Similar to:

https://github.com/liferay-frontend/liferay-portal/pull/2135#issuecomment-1104445300

Susan exhausted all other known options to indirectly wait for the portlet to appear. At surface level to me, this could be similar issue but we wouldn't truly know until we try some of these out. Could you check some of these options? If they don't work then we can go ahead with Refresh() if that's the only thing that seems to work.

**Comment by mateusralv** [comment]
ci:test:frontend:nocompile

**Comment by mateusralv** [comment]
Hi @john-co I tried the macros:

1. WaitForElementPresent() [FAIL](https://testray.liferay.com/home/-/testray/case_results/1792048188?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D1792044388%26testrayRunId%3D1792044440%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc)
2. WaitForPageLoad() [PASS](https://testray.liferay.com/home/-/testray/case_results/1792050542?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D1792044388%26testrayRunId%3D1792044440%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc)
3. LiferayEvent.waitForPortletReady() [PASS](https://testray.liferay.com/home/-/testray/case_results/1792051974?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D1792044388%26testrayRunId%3D1792044440%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc)
4. WaitForLiferayEvent.initializeLiferayEventLog() [PASS](https://testray.liferay.com/home/-/testray/case_results/1792046048?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D1792044388%26testrayRunId%3D1792044440%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc)
5. WaitForLiferayEvent(attributeName = "type", attributeValue = "allPortletsReady", eventName = "allPortletsReady") [PASS](https://testray.liferay.com/home/-/testray/case_results/1792049374?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D1792044388%26testrayRunId%3D1792044440%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc)
6. Pausing for 5sec [PASS](https://testray.liferay.com/home/-/testray/case_results/1792047004?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D1792044388%26testrayRunId%3D1792044440%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc)

All I added the macro at the beginning of the test. Of these macros which would make the test more stable? I thought of putting `WaitForPageLoad() `because it guarantees full page loading. What do you think?

**Comment by john-co** [comment]
Thanks @mateusralv! Agree with WaitForPageLoad() sounds more stable. 

cc/ @john-co testing

**Comment by john-co** [comment]
cc/ @john-co testing for slack integration

**Comment by john-co** [comment]
Upon further inspection on the flaky test, it looks like that the page loads but the portlet itself doesn't render for some reason:
image



This is different from your trial with WaitForElementPresent() [FAIL](https://testray.liferay.com/home/-/testray/case_results/1792048188?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D1792044388%26testrayRunId%3D1792044440%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc)

It looks like an improper path locator on your trial. :thinking: 
https://testray.liferay.com/reports/production/logs/2022-05/test-1-31/test-portal-acceptance-pullrequest(master)/3384/functional-tomcat90-mysql57-jdk8/0/4/LocalFile.BalloonEditor_CanFormatTextWaitForElementPresent/index.html.gz?authuser=0

So in this case, probably Refresh() will have the best chance to try and stabilize. This issue occurs 1% of the time so it's hard to reproduce.

**Comment by john-co** [comment]
So let's just use Refresh() like you had it originally. Sorry for the back and forth.

File: portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/wysiwyg/BalloonEditor.testcase [hunk 1]
```diff
@@ -26,6 +26,8 @@ definition {
 			widgetName = "CKEditor Sample");
 
 		Navigator.gotoPage(pageName = "${layoutName}");
+
+		Refresh();
```
Comment (john-co) [inline portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/wysiwyg/BalloonEditor.testcase:30]: Oh...you still left in the Refresh() here.

File: portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/wysiwyg/BalloonEditor.testcase [hunk 2]
```diff
@@ -259,6 +261,8 @@ definition {
 		var paragraphText = "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Nunc id cursus metus aliquam eleifend mi in nulla. Quam adipiscing vitae proin sagittis nisl rhoncus. Suspendisse faucibus interdum posuere lorem. Nullam ac tortor vitae purus faucibus ornare. Ac felis donec et odio pellentesque diam. Nulla at volutpat diam ut. Posuere urna nec tincidunt praesent semper feugiat nibh. Gravida quis blandit turpis cursus. Proin libero nunc consequat interdum varius. Sollicitudin ac orci phasellus egestas tellus rutrum tellus pellentesque. Neque volutpat ac tincidunt vitae semper quis lectus nulla at. Odio euismod lacinia at quis risus sed vulputate odio ut. Augue lacus viverra vitae congue eu consequat ac. Elementum sagittis vitae et leo duis ut diam. Diam quis enim lobortis scelerisque fermentum dui faucibus.";
 
 		task ("Check that text can be bolded") {
+			WaitForPageLoad();
```
Comment (john-co) [inline portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/wysiwyg/BalloonEditor.testcase:264]: We'll just drop this.

**Review by john-co — CHANGES_REQUESTED** [review]

**Review by john-co — APPROVED** [review]
made the changes, will forward.

**Comment by john-co** [comment]
ci:forward

**Comment by john-co** [comment]
cc/ @john-co testing please ignore

**Comment by john-co** [comment]
@john-co test 2

**Pull request closed without merging**